### PR TITLE
Add rhoCorr and energy-binned cuts

### PR DESCRIPTION
**Pull request by afiqaize**
[Take 2 of the closed  #18558]

This PR adds 2 new features to the affected filters:
1 - ability to do rho correction on the variable being cut on, so that this can be applied at filter level instead of producer level as is done now
2 - energy or Et dependent cut thresholds for e.g. looser dPhi cut at low Et

The customization function provided modifies the affected filters such that they're compatible with this PR but with no change in behavior at all.

Cheers,
Afiq

### CODE REVIEW

**Comment by Martin-Grunewald** [comment]
This looks the wrong strategy:
You need to modify several/all filters, as opposed to just adding it to a single/few producers,
or write a new producer which takes the collection and slims it accordingly.

**Comment by afiqaize** [comment]
The idea is to have these capabilities with as little migration work as possible, given the timescale we have - there are just many HLTEgammaGenericFilters in the menu. I'm not sure I understand the last part about the producer - currently all the relevant producers also have the same bit of duplicate code - so this is just a matter of shifting where this correction is done (without creating another collection, not sure if this would be cheaper).

What do you think is a more efficient migration strategy?

**Comment by Martin-Grunewald** [comment]
Code duplication is never a good idea.
Also, how many different working points to you foresee?
It should be restricted to a very few (for ease of efficiency / bg and rate calculations), and then it is
better to put those into the collection producer, rather than allow each and every filter instance used in the menu to do things differently.

**Comment by afiqaize** [comment]
For the upcoming single ele retuning we foresee 2 working points; Loose and Tight.

Let me try to clarify the motivation behind this PR. Currently, we have 2 seeded H/E producers in the menu: with and without rhoCorr. The latter one serves only the single electron paths. As to apply rhoCorr is part of a WP, with this PR we can have only single ele WPs apply the rhoCorr on H/E and others don't, without branching off on the producer, which we think is the more consistent approach (and less chance for definition-affecting desync).

Similarly, last year with our effArea updates we more or less "forced" other WPs to adopt them, due to not wanting to branch off the isolation producers.

If this PR is accepted, the plan is to:
1- Converge back on the H/E producer mentioned above
2- Turn off the corrections at producer level and on at filter level for Ecal and Hcal isolations
3- The retuned single ele WPs can then be applied without affecting any other WPs whatsoever

File: HLTrigger/Egamma/src/HLTEgammaGenericQuadraticEtaFilter.cc [hunk 1]
```diff
@@ -113,64 +158,116 @@ HLTEgammaGenericQuadraticEtaFilter::hltFilter(edm::Event& iEvent, const edm::Eve
...
+
+    double cutRegularEB1_ = 9999., cutRegularEE1_ = 9999.;
+    double cutRegularEB2_ = 9999., cutRegularEE2_ = 9999.;
+    double cutOverEEB1_ = 9999., cutOverEEE1_ = 9999.;
+    double cutOverEEB2_ = 9999., cutOverEEE2_ = 9999.;
+    double cutOverE2EB1_ = 9999., cutOverE2EE1_ = 9999.;
+    double cutOverE2EB2_ = 9999., cutOverE2EE2_ = 9999.;
+
+    int iEn = -1;
+    for (int dIt = energyLowEdges_.size() - 1; dIt > -1; dIt--) {
```
Comment (fwyzard) [inline HLTrigger/Egamma/src/HLTEgammaGenericQuadraticEtaFilter.cc]: please consider replacing this loop with something like `std::lower_bound` or `std::upper_bound(energyLowEdges_.begin(), energyLowEdges_.end(), energy)`

File: HLTrigger/Egamma/src/HLTEgammaGenericQuadraticFilter.cc [hunk 2]
```diff
@@ -93,28 +135,72 @@ HLTEgammaGenericQuadraticFilter::hltFilter(edm::Event& iEvent, const edm::EventS
...
+    if (energy < 0.) energy = 0.; /* first and second order terms assume non-negative energies */
+
+    // Pick the right cut threshold
+    double cutRegularEB_ = 9999., cutRegularEE_ = 9999.;
+    double cutOverEEB_ = 9999., cutOverEEE_ = 9999.;
+    double cutOverE2EB_ = 9999., cutOverE2EE_ = 9999.;
+
+    int iEn = -1;
+    for (int dIt = energyLowEdges_.size() - 1; dIt > -1; dIt--) {
+      if ( energy >= energyLowEdges_.at(dIt) ) {
```
Comment (fwyzard) [inline HLTrigger/Egamma/src/HLTEgammaGenericQuadraticFilter.cc]: please consider replacing this loop with something like `std::lower_bound` or `std::upper_bound(energyLowEdges_.begin(), energyLowEdges_.end(), energy)`

File: HLTrigger/Egamma/src/HLTGenericFilter.cc [hunk 3]
```diff
@@ -116,53 +156,97 @@ HLTGenericFilter<T1>::hltFilter(edm::Event& iEvent, const edm::EventSetup& iSetu
...
-        
+        //if (energy < 0.) energy = 0.; // require energy to be positive (needed?)
+
+        // Pick the right cut threshold
+        double cutRegularEB_ = 9999., cutRegularEE_ = 9999.;
+        double cutOverEEB_ = 9999., cutOverEEE_ = 9999.;
+        double cutOverE2EB_ = 9999., cutOverE2EE_ = 9999.;
+
+        int iEn = -1;
+        for (int dIt = energyLowEdges_.size() - 1; dIt > -1; dIt--) {
```
Comment (fwyzard) [inline HLTrigger/Egamma/src/HLTGenericFilter.cc]: please consider replacing this loop with something like `std::lower_bound` or `std::upper_bound(energyLowEdges_.begin(), energyLowEdges_.end(), energy)`

Comment (afiqaize) [inline HLTrigger/Egamma/src/HLTGenericFilter.cc]: Right, thanks for pointing it out.

**Comment by Martin-Grunewald** [comment]
This PR has now become a 92X PR - please make a PR for 91X as well.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by afiqaize** [comment]
@Martin-Grunewald not sure I understand - the milestone has both 92X and 91X in it, no?

**Comment by Martin-Grunewald** [comment]
No: this PR appears only in the 92 milestone, not the 91 milestone.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 4]
```diff
@@ -86,6 +165,18 @@ def customiseFor18429(process):
 
      return process
 
+# Add new parameters required by RecoTauBuilderConePlugin
+def customiseFor18429(process):
```
Comment (davidlange6) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py]: Hi @afiqaize - seems a duplicate definition of customiseFor18429 has been picked up. Please remove.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**